commit 6414d7aaa8ba6261b3eeae6601804216005c183b
Author: Thomas Huth <thuth@redhat.com>
Date:   Tue Jan 24 07:28:41 2017 +0100

    ibmveth: Add a proper check for the availability of the checksum features
    
    
    [ Upstream commit 23d28a859fb847fd7fcfbd31acb3b160abb5d6ae ]
    
    When using the ibmveth driver in a KVM/QEMU based VM, it currently
    always prints out a scary error message like this when it is started:
    
     ibmveth 71000003 (unregistered net_device): unable to change
     checksum offload settings. 1 rc=-2 ret_attr=71000003
    
    This happens because the driver always tries to enable the checksum
    offloading without checking for the availability of this feature first.
    QEMU does not support checksum offloading for the spapr-vlan device,
    thus we always get the error message here.
    According to the LoPAPR specification, the "ibm,illan-options" property
    of the corresponding device tree node should be checked first to see
    whether the H_ILLAN_ATTRIUBTES hypercall and thus the checksum offloading
    feature is available. Thus let's do this in the ibmveth driver, too, so
    that the error message is really only limited to cases where something
    goes wrong, and does not occur if the feature is just missing.
    
    Signed-off-by: Thomas Huth <thuth@redhat.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>
    Signed-off-by: Sasha Levin <alexander.levin@verizon.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

diff --git a/drivers/net/ethernet/ibm/ibmveth.c b/drivers/net/ethernet/ibm/ibmveth.c
index 03dca73..b375ae9 100644
--- a/drivers/net/ethernet/ibm/ibmveth.c
+++ b/drivers/net/ethernet/ibm/ibmveth.c
@@ -1604,8 +1604,11 @@ static int ibmveth_probe(struct vio_dev *dev, const struct vio_device_id *id)
 	netdev->netdev_ops = &ibmveth_netdev_ops;
 	netdev->ethtool_ops = &netdev_ethtool_ops;
 	SET_NETDEV_DEV(netdev, &dev->dev);
-	netdev->hw_features = NETIF_F_SG | NETIF_F_RXCSUM |
-		NETIF_F_IP_CSUM | NETIF_F_IPV6_CSUM;
+	netdev->hw_features = NETIF_F_SG;
+	if (vio_get_attribute(dev, "ibm,illan-options", NULL) != NULL) {
+		netdev->hw_features |= NETIF_F_IP_CSUM | NETIF_F_IPV6_CSUM |
+				       NETIF_F_RXCSUM;
+	}
 
 	netdev->features |= netdev->hw_features;
 
